rest/route/host_spawn.go

```diff
@@ -1128,53 +1129,75 @@ func (h *hostRunScript) Factory() gimlet.RouteHandler {
 
 func (h *hostRunScript) Parse(ctx context.Context, r *http.Request) error {
 	var err error
-	h.hostID, err = validateID(gimlet.GetVars(r)["host_id"])
-	if err != nil {
-		return errors.Wrap(err, "invalid host id")
-	}
-
-	hostScript := model.APIHostScript{}
-	if err = util.ReadJSONInto(util.NewRequestReader(r), &hostScript); err != nil {
+	hostScriptOpts := model.APIHostScript{}
+	if err = util.ReadJSONInto(util.NewRequestReader(r), &hostScriptOpts); err != nil {
 		return errors.Wrap(err, "can't read host command from json")
 	}
-	h.script = hostScript.Script
+	h.script = hostScriptOpts.Script
+	h.hostIDs = hostScriptOpts.Hosts
 
 	return nil
 }
 
 func (h *hostRunScript) Run(ctx context.Context) gimlet.Responder {
-	user := MustHaveUser(ctx)
-	host, err := h.sc.FindHostByIdWithOwner(h.hostID, user)
-	if err != nil {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "can't get host"))
-	}
-	if host.Status != evergreen.HostRunning {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Errorf("can't run script on host with status '%s'", host.Status))
-	}
+	u := MustHaveUser(ctx)
+	type hostOutput struct {
+		hostID string
+		output string
+	}
+	hostsOutput := make(chan hostOutput, len(h.hostIDs))
+	wg := &sync.WaitGroup{}
+	wg.Add(len(h.hostIDs))
+	for _, hostID := range h.hostIDs {
```

Comment:
Would it make sense to create a child context of `Run`'s `ctx`? That way you could cancel that context early and handle the errors.

There is an additional complication here too. IIRC `Run`'s `ctx` is canceled when the HTTP request ends, which sets a limit to how long this command can run. I'm not sure about the design of this, or what kinds of commands we plan to run, but this limits the commands to the application HTTP timeout. Is this a problem for our use case?

Reply:
I've implemented @tychoish's suggestion to return a list of process ids and poll until they're finished.